### PR TITLE
[Core] Missing strategies create

### PR DESCRIPTION
**Pull request by loumalouomega**
Part of #3185 

This adds the Create method to strategies, so #3185 can be simplified without needing to create auxiliar factories.

The main issue here is that I needed to change the ModelPart of the base SolvingStrategy to a pointer in order to have an empty constructor

### CODE REVIEW

**Comment by loumalouomega** [comment]
@philbucher ?

**Comment by loumalouomega** [comment]
Ping

File: kratos/solving_strategies/strategies/solving_strategy.h [hunk 1]
```diff
@@ -467,7 +489,7 @@ class SolvingStrategy
     ///@name Member Variables
     ///@{
 
-    ModelPart& mrModelPart;
+    ModelPart* mpModelPart;
```
Comment (philbucher) [inline kratos/solving_strategies/strategies/solving_strategy.h]: this is a delicate change

I think this is one of the reasons @pooyan-dadvand suggested to use variadic templates to pass different number of arguments to the `Create`

Comment (pooyan-dadvand) [reply kratos/solving_strategies/strategies/solving_strategy.h]: It is ok as long as we are only using the `GetModelPart` method.

File: kratos/solving_strategies/strategies/solving_strategy.h [hunk 2]
```diff
@@ -339,11 +364,11 @@ class SolvingStrategy
 
     /**
      * @brief Operations to get the pointer to the model
-     * @return mrModelPart: The model part member variable
+     * @return mpModelPart: The model part member variable
      */
     inline ModelPart& GetModelPart()
     {
-        return mrModelPart;
+        return *mpModelPart;
```
Comment (pooyan-dadvand) [inline kratos/solving_strategies/strategies/solving_strategy.h:367]: I would add a debug check if it is null

**Comment by pooyan-dadvand** [comment]
@loumalouomega Do we still need this changes?

**Comment by loumalouomega** [comment]
It is part of the PR for the factories, yes

**Comment by loumalouomega** [comment]
Part of #3185

File: kratos/solving_strategies/strategies/solving_strategy.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/solving_strategies/strategies/solving_strategy.h:367]: Of course

Comment (loumalouomega) [reply kratos/solving_strategies/strategies/solving_strategy.h:367]: Done

File: kratos/solving_strategies/strategies/solving_strategy.h [hunk 3]
```diff
@@ -339,11 +364,12 @@ class SolvingStrategy
 
     /**
      * @brief Operations to get the pointer to the model
-     * @return mrModelPart: The model part member variable
+     * @return mpModelPart: The model part member variable
      */
     inline ModelPart& GetModelPart()
     {
-        return mrModelPart;
+        KRATOS_ERROR_IF(mpModelPart == nullptr) << "ModelPart in the SolvingStrategy is not initialized" << std::endl;
```
Comment (philbucher) [inline kratos/solving_strategies/strategies/solving_strategy.h]:
```suggestion
        KRATOS_ERROR_IF_NOT(mpModelPart) << "ModelPart in the SolvingStrategy is not initialized" << std::endl;
```

File: kratos/solving_strategies/strategies/solving_strategy.h [hunk 4]
```diff
@@ -129,14 +138,17 @@ class SolvingStrategy
...
     {
         SetMoveMeshFlag(MoveMeshFlag);
     }
 
     /** Destructor.
      */
-    virtual ~SolvingStrategy(){}
+    virtual ~SolvingStrategy()
+    {
+        mpModelPart =  nullptr;
```
Comment (philbucher) [inline kratos/solving_strategies/strategies/solving_strategy.h]: why is this needed?

File: kratos/solving_strategies/strategies/solving_strategy.h [hunk 5]
```diff
@@ -109,13 +111,20 @@ class SolvingStrategy
     ///@name Life Cycle
     ///@{
 
+    /**
+     * @brief Default constructor
+     */
+    explicit SolvingStrategy() : mpModelPart(nullptr)
+    {
+    }
```
Comment (philbucher) [inline kratos/solving_strategies/strategies/solving_strategy.h]:
```suggestion
    explicit SolvingStrategy() { }
```

I would not initialize it if the value it has is already the default

**Review by philbucher — COMMENTED** [review]
minor otherwise ok from my side

File: kratos/solving_strategies/strategies/solving_strategy.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/solving_strategies/strategies/solving_strategy.h]: I don't trust compilers and its decisions, better make it explicit

File: kratos/solving_strategies/strategies/solving_strategy.h (hunk 5 above)
Comment (loumalouomega) [reply kratos/solving_strategies/strategies/solving_strategy.h]: Idem, I don't trust compilers decissions. This is harmless

File: kratos/solving_strategies/strategies/solving_strategy.h (hunk 4 above)
Comment (philbucher) [reply kratos/solving_strategies/strategies/solving_strategy.h]: but it confuses others
This makes me think why is this necessary. Is there some special memory-related treatment necessary? (I know it is not but others reading might not now)

File: kratos/solving_strategies/strategies/solving_strategy.h (hunk 5 above)
Comment (philbucher) [reply kratos/solving_strategies/strategies/solving_strategy.h]: what do you mean? The member is initialized

File: kratos/solving_strategies/strategies/solving_strategy.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/solving_strategies/strategies/solving_strategy.h]: I know it works if I remove it, but my faith in compilers is being shrinking with the time

Comment (philbucher) [reply kratos/solving_strategies/strategies/solving_strategy.h]: can you explain your concerns? I have not faced such problems

Comment (loumalouomega) [reply kratos/solving_strategies/strategies/solving_strategy.h]: I prefer to do obvious code than ambiguous code that I know that one day the compiler will not compile due to criteria change. I am just compiler nihilist

Comment (pooyan-dadvand) [reply kratos/solving_strategies/strategies/solving_strategy.h]: Is not needed because after the destructor call you are the `mpModelPart` member variable. So the value you assigned to it wouldn't affect anywhere.

Comment (pooyan-dadvand) [reply kratos/solving_strategies/strategies/solving_strategy.h]: However, it is very important to set the mpModelPart to `nullptr` in the constructor if it is not initialized to avoid uninitialized pointer.

Comment (loumalouomega) [reply kratos/solving_strategies/strategies/solving_strategy.h]: Understood, I have already deleted it

Comment (loumalouomega) [reply kratos/solving_strategies/strategies/solving_strategy.h]: > However, it is very important to set the mpModelPart to `nullptr` in the constructor if it is not initialized to avoid uninitialized pointer.

Reverting the commit then

Comment (loumalouomega) [reply kratos/solving_strategies/strategies/solving_strategy.h]: It is already initialized in the variable declaration

Comment (loumalouomega) [reply kratos/solving_strategies/strategies/solving_strategy.h]: Can we proceed then? @pooyan-dadvand @philbucher

**Review by philbucher — DISMISSED** [review]
:+1:

**Comment by loumalouomega** [comment]
Sorry I needed to solve a conflict

**Review by philbucher — DISMISSED** [review]

**Comment by loumalouomega** [comment]
Sorry @philbucher the conflict solve of Github duplicated some code, I needed to fix it

**Review by philbucher — APPROVED** [review]

**Comment by philbucher** [comment]
I checked #3185 and I couldn't see why the default constructors are needed

Can you explain please and wait with merging?
I think having those can be dangerous since the internals are not initialized

**Comment by loumalouomega** [comment]
> I checked #3185 and I couldn't see why the default constructors are needed
> 
> Can you explain please and wait with merging?
> I think having those can be dangerous since the internals are not initialized

The default constructors are needed in order to add the class to the KratosComponents

**Comment by philbucher** [comment]
Ok makes sense
Go ahead then please

**Comment by pooyan-dadvand** [comment]
> > I checked #3185 and I couldn't see why the default constructors are needed
> > Can you explain please and wait with merging?
> > I think having those can be dangerous since the internals are not initialized
> 
> The default constructors are needed in order to add the class to the KratosComponents

If the reason is the adding to Kratos components then I think it is avoidable. But they are needed for serialization anyway.

File: kratos/solving_strategies/strategies/solving_strategy.h [hunk 6]
```diff
@@ -464,7 +485,7 @@ class SolvingStrategy
     ///@name Member Variables
     ///@{
 
-    ModelPart& mrModelPart;
+    ModelPart* mpModelPart = nullptr;
```
Comment (philbucher) [inline kratos/solving_strategies/strategies/solving_strategy.h:488]: I recently read about [`std::ref`](https://en.cppreference.com/w/cpp/utility/functional/ref), I think this could have also be used here

Comment (loumalouomega) [reply kratos/solving_strategies/strategies/solving_strategy.h:488]: The problem is that I need to be able to define an empty constructor, and therefore it is not possible to define a reference

Comment (philbucher) [reply kratos/solving_strategies/strategies/solving_strategy.h:488]: good point
I was thinking that std::ref can be reassigned, but not that is cannot be empty
